truffle/src/main/java/org/jruby/truffle/core/string/StringNodes.java

```diff
@@ -1355,6 +1365,307 @@ private int prevCharHead(Encoding enc, byte[]bytes, int p, int s, int end) {
+
+        private static final byte[] SCRUB_REPL_UTF8 = new byte[]{(byte)0xEF, (byte)0xBF, (byte)0xBD};
+        private static final byte[] SCRUB_REPL_ASCII = new byte[]{(byte)'?'};
+        private static final byte[] SCRUB_REPL_UTF16BE = new byte[]{(byte)0xFF, (byte)0xFD};
+        private static final byte[] SCRUB_REPL_UTF16LE = new byte[]{(byte)0xFD, (byte)0xFF};
+        private static final byte[] SCRUB_REPL_UTF32BE = new byte[]{(byte)0x00, (byte)0x00, (byte)0xFF, (byte)0xFD};
+        private static final byte[] SCRUB_REPL_UTF32LE = new byte[]{(byte)0xFD, (byte)0xFF, (byte)0x00, (byte)0x00};
+
+
+        @Child private IsTaintedNode isTaintedNode = IsTaintedNode.create();
+        @Child private TaintNode taintNode = TaintNode.create();
+        @Child private CallDispatchHeadNode strCompatAndValidNode = DispatchHeadNodeFactory.createMethodCall();
+        @Child private RopeNodes.MakeConcatNode makeConcatNode = RopeNodesFactory.MakeConcatNodeGen.create(null, null, null);
+
+        @Specialization
+        public DynamicObject scrubNoBlock(VirtualFrame frame, DynamicObject string, Object repl, NotProvided block) {
+            return scrubDefault(frame, string, repl, nil());
+        }
+
+        @Specialization
+        public DynamicObject scrubDefault(VirtualFrame frame, DynamicObject string, Object repl, DynamicObject block) {
+            final DynamicObject rubyEncoding = getContext().getEncodingManager().getRubyEncoding(StringOperations.encoding(string));
+            final Rope rope = rope(string);
+            final ByteList value = RopeOperations.getByteListReadOnly(rope);
+
+            CodeRange cr = rope.getCodeRange();
+            Encoding enc;
+            Encoding encidx;
+            Rope buf = null;
+            byte[] repBytes;
+            int rep;
+            int replen;
+            boolean tainted = false;
+
+            if (cr == CR_7BIT || cr == CR_VALID)
+                return nil();
+
+            enc = rope.getEncoding();
+            if (repl != nil()) {
+                repl = strCompatAndValidNode.call(frame, string, "str_compat_and_valid", repl, rubyEncoding);
+                tainted |= isTaintedNode.executeIsTainted(repl);
+            }
+
+            if (enc.isDummy()) {
+                return nil();
+            }
+            encidx = enc;
+
+            if (enc.isAsciiCompatible()) {
+                byte[] pBytes = value.unsafeBytes();
+                int p = value.begin();
+                int e = p + value.getRealSize();
+                int p1 = p;
+                boolean rep7bit_p;
+                if (block != nil()) {
+                    repBytes = null;
+                    rep = 0;
+                    replen = 0;
+                    rep7bit_p = false;
+                }
+                else if (repl != nil()) {
+                    final Rope replRope = rope((DynamicObject) repl);
+                    final ByteList replValue = RopeOperations.getByteListReadOnly(replRope);
```

Comment:
@nirvdrum I tried it again and it appears to work now. I just pushed a new commit